### PR TITLE
Use pagination parameters from SerpApi instead of calculating on the client

### LINKED ISSUE #22: Pagination iterator doesn't work for APIs with token-based pagination
For several APIs, parsing the `serpapi_pagination.next` is the only way to update `params_dict` with correct values. [An increment of  `params.start`](https://github.com/serpapi/google-search-results-python/blob/ed7797c132d80613080b11b99f5b137bbeb5c3f5/serpapi/pagination.py#L26-L27) won't work for Google Scholar Profiles, Google Maps, YouTube.

https://github.com/serpapi/google-search-results-python/blob/ed7797c132d80613080b11b99f5b137bbeb5c3f5/serpapi/pagination.py#L26-L27

**Google Scholar Profiles**

[Google Scholar Profiles API](https://serpapi.com/google-scholar-profiles-api) have `pagination.next_page_token` instead of `serpapi_pagination.next`.

`pagination.next` is a next page URI like `https://serpapi.com/search.json?after_author=0QICAGE___8J&engine=google_scholar_profiles&hl=en&mauthors=label%3Asecurity` where `after_author` is set to `next_page_token`.

**Google Maps**

In [Google Maps Local Results API](https://serpapi.com/maps-local-results) there's only `serpapi_pagination.next` with a URI like `https://serpapi.com/search.json?engine=google_maps&ll=%4040.7455096%2C-74.0083012%2C14z&q=Coffee&start=20&type=search`

**YouTube**

In [YouTube Search Engine Results API](https://serpapi.com/youtube-search-api) there's `serpapi_pagination.next_page_token` similar to Google Scholar Profiles. `serpapi_pagination.next` is a URI with `sp` parameter set to `next_page_token`.

@jvmvik What do you think about parsing `serpapi_pagination.next` in `Pagination#__next__`?

```diff
- self.start += self.page_size
+ self.client.params_dict.update(dict(parse.parse_qsl(parse.urlsplit(result['serpapi_pagination']['next']).query)))
```

[Here's an example](https://replit.com/@serpapi/scrape-google-scholar-authors-python-serpapi) of endless pagination of Google Scholar Authors (scraped 190 pages and manually stopped).

### LINKED ISSUE #25: google scholar pagination not returning final results page
I am using the paginate method with google scholar engine to return all results for a search term. When I use a for loop to iterate the pagination and put the results a list, it doesn't return the final page of results, instead stopping at the penultimate page (code snippet and terminal output below).

```python
import serpapi
import os
from loguru import logger
from dotenv import load_dotenv

load_dotenv()

search_string = '"Singer Instruments" PhenoBooth'

# Pagination allows iterating through all pages of results
logger.info("Initialising search through serpapi")
search = serpapi.GoogleSearch(
    {
        "engine": "google_scholar",
        "q": search_string,
        "api_key": os.getenv("SERPAPI_KEY"),
        "as_ylo": 1900,
    }
)
pages = search.pagination(start=0, page_size=20)

# get dict for each page of results and store in list
results_list = []
page_number = 1
for page in pages:
    logger.info(f"Retrieving results page {page_number}")
    results_list.append(page)
    page_number += 1

gscholar_results = results_list[0]["search_information"]["total_results"]
print(f"results reported by google scholar: {gscholar_results}")

paper_count = 0
for page in results_list:
    for paper in page["organic_results"]:
        paper_count += 1

print(f"number of papers in results: {paper_count}")
```

Screenshot 2021-07-30 at 17 05 11

If I check my searches on serpAPI.com, results are being generated for all pages (see below for example in code). So the problem is not that the result isn't generated, its just not coming out of the pagination iterator for some reason.

Screenshot 2021-07-30 at 17 06 21

### LINKED ISSUE #26: google scholar pagination skips result 20
When retrieving results from Google Scholar using the pagination() method, the first article on the second page of google scholar is always missing. 

I think this is caused by the following snippet in the update() method of google-search-results-python/serpapi/pagination.py:

```python
def update(self):
        self.client.params_dict["start"] = self.start
        self.client.params_dict["num"] = self.num
        if self.start > 0:
            self.client.params_dict["start"] += 1
```

This seems to mean that for all pages except the first, paginate increases start by 1. So while for the first page it requests results starting at 0 and ending at 19 (if page_size=20). For the second page it requests results starting at 21 and ending at 40, skipping result 20.

If I delete the if statement, the code seems to work as intended and I get result 19 back.

### LINKED ISSUE #32: [Google Jobs API] Support for Pagination
As Google Jobs does not return `serpapi_pagination` key but expects `start` param to paginate, this iteration of the library does not support pagination in Google Jobs. Pagination Support to be added for Google Jobs.

```
# stop if backend miss to return serpapi_pagination
if not 'serpapi_pagination' in result:
  raise StopIteration

# stop if no next page
if not 'next' in result['serpapi_pagination']:
    raise StopIteration
```

image

### LINKED ISSUE #38: [Pagination] Pagination isn't correct and it skips index by one
image

Since the `start` value starts from `0`, the correct second page should be `10` and not `11`.

This behaviour is causing a skip in pages also. The customers are getting confusing results:
image

[Intercom Link](https://app.intercom.com/a/apps/yzck75jr/inbox/inbox/conversation/102910000035805)
First recognized by @marm123.

I think this part needs to be replaced by:
image

```py
self.client.params_dict['start'] += 0
```

Whether it would cause any error on other engines is something I don't know. But it may also fix it for every other engine.

### PR DESCRIPTION
**Pull request by ilyazub**
`start` and `num` parameters are not suitable for token-based pagination. Such pagination is used on Google Maps, YouTube, Google Scholar Authors, and other search engines.

This PR consumes URL query parameters for the next page. It stops paginating when parameters do not change.

Details: https://github.com/serpapi/google-search-results-python/issues/22

Some tests are failing because `start` and `num` parameters are not supported anymore. These tests will be fixed in the following commits.

---

Fixes #38 
Fixes #32 
Fixes #26 
Fixes #25
Fixes #22

### CODE REVIEW

File: tests/test_duck_duck_go_search.py [hunk 1]
```diff
@@ -23,5 +23,46 @@ def test_get_json(self):
 				# pp.pprint(data)
 				self.assertTrue(len(data.keys()) > 3)
 
+		@unittest.skipIf((os.getenv("API_KEY") == None), "no api_key provided")
+		def test_paginate_page_size(self):
```
Comment (ilyazub) [inline tests/test_duck_duck_go_search.py:27]: @dimitryzub, can you help me with tests based on pagination examples you've told?

File: tests/test_ebay_search.py [hunk 2]
```diff
@@ -21,22 +21,30 @@ def test_get_json(self):
 				# pp = pprint.PrettyPrinter(indent=2)
 				# pp.pprint(data)
 
-    # TODO fix universal pagination
-		@unittest.skipIf((os.getenv("DEBUGAPI_KEY") == None), "no api_key provided")
+		@unittest.skipIf((os.getenv("API_KEY") == None), "no api_key provided")
```
Comment (ilyazub) [inline tests/test_ebay_search.py:24]: @dimitryzub here's another example test case for your test scenarios.

File: tests/test_duck_duck_go_search.py (hunk 1 above)
Comment (dimitryzub) [reply tests/test_duck_duck_go_search.py:27]: @ilyazub I've used the `pagination()` method which is worked without any workarounds.

Do we need to check for title equality as this is for test purposes to show if there're any duplicates? Or for something else?

https://github.com/serpapi/google-search-results-python/blob/bd65ec74a840f8fbe0b5f2c33512a3468009a5f6/tests/test_duck_duck_go_search.py#L56-L57

DuckDuckGo tends to display the same thing over multiple pages so there's could be a lot of duplicates. In my test where I paginate over all pages and extracted titles, the duplicate number was varying between 2 and 16.

____

`count % 2, 0` causes confusion to me. 

https://github.com/serpapi/google-search-results-python/blob/bd65ec74a840f8fbe0b5f2c33512a3468009a5f6/tests/test_duck_duck_go_search.py#L61

Sorry for not particularly useful comment.

File: tests/test_ebay_search.py (hunk 2 above)
Comment (dimitryzub) [reply tests/test_ebay_search.py:24]: @ilyazub Currently, I'm not really sure what else to add here. 

The issue I had with eBay is I had to check if the current page == next page, and if equal - exit the loop i.e `raise StopIteration`, which is done by checking:

https://github.com/serpapi/google-search-results-python/blob/b733a1a4375c8393e8242ae745f9b6c75c0d42f6/serpapi/pagination.py#L57-L58

If I'm not getting it wrong, it should work correctly.

**Review by ilyazub — COMMENTED** [review]
Tests [fail](https://github.com/serpapi/google-search-results-python/runs/6694602087?check_suite_focus=true) because [Walmart Search API pagination is broken](https://github.com/serpapi/public-roadmap/issues/163), [Google Local Results API are broken](https://github.com/serpapi/public-roadmap/issues/130), and several tests are flaky.
